Announce automatic checking of cfgs at compile-time (#1313)
rust-lang/blog.rust-lang.org: posts/2024-04-29-check-cfg.md

---
layout: post
title: "Automatic checking of cfgs at compile-time"
author: Urgau
team: The Cargo Team <https://www.rust-lang.org/governance/teams/dev-tools#cargo>
---

# Automatic checking of cfgs at compile-time

> The Cargo and Compiler team are delighted to announce that starting with Rust 1.80 (or nightly-2024-04-XX) every _reachable_ `#[cfg]`s will be automatically checked for expected config names and values.

Reviewer: ```suggestion
The Cargo and Compiler team are delighted to announce that starting with Rust 1.80 (or nightly-2024-04-XX) every _reachable_ `#[cfg]` will be automatically checked that they match the expected config names and values.
```

This can help with verifying that the crate is correctly handling conditional compilation for different target platforms or features. It ensures that the cfg settings are consistent between what is intended and what is used, helping to catch potential bugs or errors early in the development process.

> Addressing a common pitfall for new and advanced users.

Reviewer: ```suggestion
This addresses a common pitfall for new and advanced users.
```

> This is another step to our commitment to provide user-focused tooling and we are eager and excited to finally see it fixed, more than two years since the original [RFC 3013](https://github.com/rust-lang/rfcs/pull/3013)[^1].

Reviewer: ```suggestion
This is another step to our commitment to provide user-focused tooling and we are eager and excited to finally see it fixed, after more than two years since the original [RFC 3013](https://github.com/rust-lang/rfcs/pull/3013)[^1].
```

[^1]: The stabilized implementation and RFC 3013 diverge significantly, in particular there is only one form for `--check-cfg`: `cfg()` (instead of `values()` and `names()` being incomplete and subtlety incompatible with each other).

## A look at the feature 

> Every time a Cargo feature is declared that feature is transformed into a config that is passed to `rustc` (the Rust compiler) so it can verify with its [well known cfgs](TODO) if any of the `#[cfg]`, `#![cfg_attr]` and `cfg!` have unexpected configs and report a warning with the `unexpected_cfgs` lint.

Author: ```suggestion
Every time a Cargo feature is declared that feature is transformed into a config that is passed to `rustc` (the Rust compiler) so it can verify with its [well known cfgs](https://doc.rust-lang.org/nightly/rustc/check-cfg.html#well-known-names-and-values) if any of the `#[cfg]`, `#![cfg_attr]` and `cfg!` have unexpected configs and report a warning with the `unexpected_cfgs` lint.
```

name = "foo"

[features]
lasers = []
zapping = []
```

*`src/lib.rs`:*
```rust
#[cfg(feature = "lasers")]  // This condition is expected
                            //   as "lasers" is an expected value of `feature`
fn shoot_lasers() {}

#[cfg(feature = "monkeys")] // This condition is UNEXPECTED
                            //   as "monkeys" is NOT an expected value of `feature`
fn write_shakespeare() {}

#[cfg(windosw)]             // This condition is UNEXPECTED
                            //   it's supposed to be `windows`
fn win() {}
```

*`cargo check`*:
![cargo-check](https://github.com/rust-lang/rust/assets/3616612/c6ecdb34-b92c-42b8-9f80-7066b76541ff)

## Custom cfgs and build scripts

> In Cargo point-of-view: a custom cfg is one that is neither defined by `rustc` nor by a Cargo feature. Think of `tokio_unstable`, `has_foo`, ... but not `feature = "lasers"`, `unix`, ...

Some crates use custom cfgs that they either expected from the environment (`RUSTFLAGS`or other means) or is enabled by some logic in the crate `build.rs`. For those crates Cargo provides a new instruction: [`cargo::rustc-check-cfg`](TODO)[^2] (or `cargo:rustc-check-cfg` for older Cargo version).

> [^2]: `cargo::rustc-check-cfg` take action since Rust 1.80, between Rust 1.77 and Rust 1.79 *(included)* it silently did nothing and before that (so Rust 1.76 and below) it warned when used without the unstable Cargo `-Zcheck-cfg`.

Reviewer: ```suggestion
[^2]: `cargo::rustc-check-cfg` started working in Rust 1.80, between Rust 1.77 and Rust 1.79 *(included)* it silently did nothing and before that (so Rust 1.76 and below) it warned when used without the unstable Cargo `-Zcheck-cfg`.
```

Author: When the blog post will be published, Rust 1.80 (or another version) will not yet be stable.

We should therefore use the future: `will start working in...`.
What do you think?

Reviewer: sounds good!

> [^2]: `cargo::rustc-check-cfg` will start working in Rust 1.80 (or nightly-2024-04-XX), between Rust 1.77 and Rust 1.79 *(included)* it silently did nothing and before that (so Rust 1.76 and below) it warned when used without the unstable Cargo `-Zcheck-cfg`.

Reviewer 2: ```suggestion
[^2]: `cargo::rustc-check-cfg` will start working in Rust 1.80 (or nightly-2024-04-XX). From Rust 1.77 to Rust 1.79 *(inclusive)* it is silently ignored. In Rust 1.76 and below a warning is emitted when used without the unstable Cargo flag `-Zcheck-cfg`.
```

> The syntax to use is described in the [rustc book](https://doc.rust-lang.org/rustc/index.html) section [checking configuration](TODO), but in a nut shell the basic syntax of `--check-cfg` is:

Reviewer: ```suggestion
The syntax to use is described in the [rustc book](https://doc.rust-lang.org/rustc/index.html) section [checking configuration](TODO), but in a nutshell the basic syntax of `--check-cfg` is:
```

> The syntax to use is described in the [rustc book](https://doc.rust-lang.org/rustc/index.html) section [checking configuration](TODO), but in a nutshell the basic syntax of `--check-cfg` is:

Author: ```suggestion
The syntax to use is described in the [rustc book](https://doc.rust-lang.org/nightly/rustc/) section [checking configuration](https://doc.rust-lang.org/nightly/rustc/check-cfg.html), but in a nutshell the basic syntax of `--check-cfg` is:
```

Note that every custom cfgs must always be expected, regardless if the cfg is active or not!

### `build.rs` example

`build.rs`:
```rust
fn main() {
    println!("cargo::rustc-check-cfg=cfg(has_foo)"); // <-- new with Cargo 1.80
    if has_foo() {
        println!("cargo::rustc-cfg=has_foo");
    }
}
```

> Note: Each `cargo::rustc-cfg` must always have a accompanying _unconditional_ `cargo::rustc-check-cfg` directive otherwise you risk having warnings like this:`unexpected cfg condition name: has_foo`.

### Equivalence table

| `cargo::rustc-cfg`      | `cargo::rustc-check-cfg`                       |
|-------------------------|------------------------------------------------|
| `foo`                   | `cfg(foo)` or `cfg(foo, values(none()))`       |
| `foo=""`                | `cfg(foo, values(""))`                         |
| `foo="bar"`             | `cfg(foo, values("bar"))`                      |
| `foo="1"` and `foo="2"` | `cfg(foo, values("1", "2"))`                   |
| `foo="1"` and `bar="2"` | `cfg(foo, values("1"))` and `cfg(bar, values("2"))` |
| `foo` and `foo="bar"`   | `cfg(foo, values(none(), "bar"))`              |

> More details can be found on the [`rustc` book](TODO).

Author: ```suggestion
More details can be found on the [`rustc` book](https://doc.rust-lang.org/nightly/rustc/check-cfg.html).
```

## Frequently asked questions

### Can it be disabled?

> No, it's **always on** for Cargo and _cannot_ be disabled.

Reviewer 3: Can't one use `#![allow(unexpected_cfgs)]` to disable it?

Author: I was talking from Cargo point-of-view, but yes, of course inside the code the lint can be controlled like any other lints.

What about adding:
```suggestion
The feature is **always on** and _cannot_ be disabled from Cargo, but like any other lints it can be controlled: `#![allow(unexpected_cfgs)]`.
```

Author: Hum, with Cargo `[lints]` table my suggestion is (still) wrong.

Maybe it's just better to remove that question entirely.

Reviewer 3: Either way, it is up to you. Your proposed change seems fine to me.

### How does it interact with the `RUSTFLAGS` env ?

You should be able to use the `RUSTFLAGS` environment variable like it was before.
*Currently `--cfg` arguments are not checked, only usage in code are.*

This means that doing `RUSTFLAGS="--cfg tokio_unstable" cargo check` will not report any warnings, unless `tokio_unstable` is used within your local crates, in which case crate author will need to make sure that that custom cfg is expected with `cargo::rustc-check-cfg` in the `build.rs` of that crate.

### How to expect custom cfgs without a `build.rs` ?

> There is not **currently no way** to expect a custom cfg other than with `cargo::rustc-check-cfg` in a `build.rs`.

Reviewer: ```suggestion
There is **currently not any way** to expect a custom cfg other than with `cargo::rustc-check-cfg` in a `build.rs`.
```

Reviewer: maybe you need a section or at least a link for how to use this when using rustc without cargo, e.g. in the Linux Kernel.

Author: This blog post is very much centered around Cargo users (it is even posted by the Cargo team). So a full section for non-Cargo users feels overkill and out of place.

However we could have question that basicly just links to the rustc docs, which contains all the details for that.

Btw, the Linux Kernel tested the feature before stabilization and is planning on using it.

Reviewer: maybe phrase it something like:

> Cargo achieves this by using the `--whatever-its-called` rustc options

which allows you to smoothly link to rustc's docs there.

Author: I just pushed a new question/section regarding non-Cargo based build system, which I think should address your remark.

Feel free to leave review comments on it.

> Crate author that don't want to use a `build.rs` are encouraged to use Cargo features instead.

Reviewer 2: ```suggestion
Crate authors that don't want to use a `build.rs` are encouraged to use Cargo features instead.
```

> ### Does it affect dependencies ?

Reviewer 2: ```suggestion
### Does the lint affect dependencies?
```
